### DISCUSSION TITLE
Failover tunnel becoming active stops routing private network

### DISCUSSION THREAD

**Discussion by elfin-sbreuers** [strongswan/strongswan#2562] [Q&A]
Our communications service provider gives us _two gateways_ to our IPsec network. Allowing us to connect via two IPsec tunnels that should route the _same_ private networks. The second gateway tunnel should act as a failover, that takes over when the other tunnel breaks down or goes into maintenance.

We configured our setup with the `trap` feature to start the second tunnel only if traffic is sent over it. While the first tunnel established takes the main load of the traffic, the second tunnel gets some traffic once in while and spins up.

But when the second tunnel gets established our `strongswan` stops routing the private network. When using `tcpdump` to track the ping of our devices to services behind our `strongswan` server, we only see the traffic reaching the public interface. After that the icmp is gone, no routing happens. When we deactivate the second child SA, keeping the IKE active, the routing starts working again and we see packages being routed.

```
// swanctl.conf
internal_network {
    start_action=trap
    rekey_time=3600s
    esp_proposals=aes128-sha1-modp1024
    remote_ts=10.240.0.0/24,100.94.242.0/24,10.54.108.0/24
    local_ts=10.20.0.0/22
    dpd_action=restart
}
connections {
    IPsec-device-vpn-tunnel1 {
        rekey_time=28800s
        version=2
        local_addrs=x.x.x.x
        remote_addrs=y.y.y.y
        proposals=aes128-sha1-modp1024
        dpd_delay=10s
        children {
            routing: internal_network {}
        }
    }
    IPsec-device-vpn-tunnel2 {
        rekey_time=28800s
        version=2
        local_addrs=x.x.x.x
        remote_addrs=z.z.z.z
        proposals=aes128-sha1-modp1024
        dpd_delay=10s
        children {
            routing: internal_network {}
        }
    }
}
```
Apart from the `swanctl.conf` everything is on default values, meaning the 220 routing table is created and automatically populated.

Searching for this kind of setup via google or in the extensive strongswan documentation did not lead me to something that helped me understand why this problem occurs. The options I found where mainly related to routing different networks. 

Do you have any pointers as to where we could look why the routing stops working? Or which configuration option should be used to allow the second tunnel standing by or not interfering with the perfectly working first tunnel?

Any help would be much appreciated.

**Comment by tobiasbrunner** [comment]
> We configured our setup with the `trap` feature to start the second tunnel only if traffic is sent over it.

You used this on both configs and there is no way to differentiate between the two (the traffic selectors are the same, so the same reqid is allocated). That is, there is no way to control whether the "second" or the "first" tunnel comes up when a packet hits the trap policy and triggers an acquire (initially or after one of them goes down).


> While the first tunnel established takes the main load of the traffic, the second tunnel gets some traffic once in while and spins up.

How would it get only "some traffic"? You mean if the first SA goes down?

> But when the second tunnel gets established our `strongswan` stops routing the private network.

That's if the first SA is still established? I don't really see why that should influence the routing. It will have an influence on the policies, though. Maybe have a look at the logs, the status output, the SAs/policies in the kernel and the routes (see #196).

You might also want to look into [route-based VPNs](https://docs.strongswan.org/docs/latest/features/routeBasedVpn.html).

**Comment by elfin-sbreuers** [comment]
Thanks for the quick response.

> How would it get only "some traffic"?

That was my interpretation, since the second tunnel comes up despite using trap. So I assumed some kind of traffic must happen, otherwise it would come online.

> You mean if the first SA goes down?

This happens while the one that started first is still up and running.

>  the traffic selectors are the same, so the same reqid is allocated

I guess that is the issue then. When the second tunnel comes on, the routing is no longer possible.

I will look into the route based VPNs approach and post a solution if I find an elegant one. Thanks again for your pointers.

**Reply by tobiasbrunner** [reply]
> This happens while the one that started first is still up and running.

Seems unlikely.


> I guess that is the issue then. When the second tunnel comes on, the routing is no longer possible.

Again, routing, not so much probably.

**Comment by elfin-sbreuers** [comment]
> Seems unlikely.

Just out of curiosity, can you image, why this happens? Because unlikely as it seems that configuration uses `trap` and eventually two tunnels are up and running.

**Reply by tobiasbrunner** [reply]
Not if "the one that started first is still up and running" is true. If that's not the case (e.g. if it went down due to a retransmission timeout), then sure, an acquire from the kernel could get triggered and that could initiate the second SA. But without logs, I really can't tell.

**Reply by elfin-sbreuers** [reply]
Thanks for your response. I used `swanctl -T` to trace the log. I replaced the public IPs of the gateways with  `y.y.y.y` and `z.z.z.z` respectively and mine with `x.x.x.x` in the logs and configuration. 

```
16[NET] received packet: from y.y.y.y[500] to x.x.x.x[500] (636 bytes)
16[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) ]
16[IKE] y.y.y.y is initiating an IKE_SA
16[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
16[IKE] remote host is behind NAT
16[ENC] generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
16[NET] sending packet: from x.x.x.x[500] to y.y.y.y[500] (344 bytes)
12[NET] received packet: from y.y.y.y[4500] to x.x.x.x[4500] (332 bytes)
12[ENC] parsed IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
12[CFG] looking for peer configs matching x.x.x.x[x.x.x.x]...y.y.y.y[y.y.y.y]
12[CFG] selected peer config 'IPsec-device-vpn-tunnel2'
12[IKE] authentication of 'y.y.y.y' with pre-shared key successful
12[IKE] authentication of 'x.x.x.x' (myself) with pre-shared key
12[IKE] IKE_SA IPsec-device-vpn-tunnel2[1] established between x.x.x.x[x.x.x.x]...y.y.y.y[y.y.y.y]
12[IKE] scheduling rekeying in 27765s
12[IKE] maximum IKE_SA lifetime 30645s
12[CFG] selected proposal: ESP:AES_CBC_128/HMAC_SHA1_96/NO_EXT_SEQ
12[IKE] CHILD_SA routing{3} established with SPIs cc3aa04c_i cbe6804b_o and TS 10.20.0.0/22 === 10.54.108.0/24 10.240.0.0/24 100.94.242.0/24
12[ENC] generating IKE_AUTH response 1 [ IDr AUTH SA TSi TSr ]
12[NET] sending packet: from x.x.x.x[4500] to y.y.y.y[4500] (236 bytes)
07[IKE] sending DPD request
07[ENC] generating INFORMATIONAL request 0 [ ]
07[NET] sending packet: from x.x.x.x[4500] to y.y.y.y[4500] (76 bytes)
02[NET] received packet: from y.y.y.y[4500] to x.x.x.x[4500] (76 bytes)
02[ENC] parsed INFORMATIONAL response 0 [ ]
07[IKE] sending DPD request
07[ENC] generating INFORMATIONAL request 1 [ ]
07[NET] sending packet: from x.x.x.x[4500] to y.y.y.y[4500] (76 bytes)
04[NET] received packet: from y.y.y.y[4500] to x.x.x.x[4500] (76 bytes)
04[ENC] parsed INFORMATIONAL response 1 [ ]
07[IKE] sending DPD request
07[ENC] generating INFORMATIONAL request 2 [ ]
07[NET] sending packet: from x.x.x.x[4500] to y.y.y.y[4500] (76 bytes)
04[NET] received packet: from y.y.y.y[4500] to x.x.x.x[4500] (76 bytes)
04[ENC] parsed INFORMATIONAL response 2 [ ]
15[NET] received packet: from z.z.z.z[500] to x.x.x.x[500] (636 bytes)
15[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) ]
15[IKE] z.z.z.z is initiating an IKE_SA
15[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
15[IKE] remote host is behind NAT
15[ENC] generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
15[NET] sending packet: from x.x.x.x[500] to z.z.z.z[500] (344 bytes)
14[NET] received packet: from z.z.z.z[4500] to x.x.x.x[4500] (332 bytes)
14[ENC] parsed IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
14[CFG] looking for peer configs matching x.x.x.x[x.x.x.x]...z.z.z.z[z.z.z.z]
14[CFG] selected peer config 'IPsec-device-vpn-tunnel1'
14[IKE] authentication of 'z.z.z.z' with pre-shared key successful
14[IKE] authentication of 'x.x.x.x' (myself) with pre-shared key
14[IKE] IKE_SA IPsec-device-vpn-tunnel1[2] established between x.x.x.x[x.x.x.x]...z.z.z.z[z.z.z.z]
14[IKE] scheduling rekeying in 26984s
14[IKE] maximum IKE_SA lifetime 29864s
14[CFG] selected proposal: ESP:AES_CBC_128/HMAC_SHA1_96/NO_EXT_SEQ
14[IKE] CHILD_SA routing{4} established with SPIs cb19c5ef_i c1c3e12e_o and TS 10.20.0.0/22 === 10.54.108.0/24 10.240.0.0/24 100.94.242.0/24
14[ENC] generating IKE_AUTH response 1 [ IDr AUTH SA TSi TSr ]
14[NET] sending packet: from x.x.x.x[4500] to z.z.z.z[4500] (236 bytes)
04[IKE] sending DPD request
04[ENC] generating INFORMATIONAL request 3 [ ]
04[NET] sending packet: from x.x.x.x[4500] to y.y.y.y[4500] (76 bytes)
08[NET] received packet: from y.y.y.y[4500] to x.x.x.x[4500] (76 bytes)
08[ENC] parsed INFORMATIONAL response 3 [ ]
```

This happens based on this configuration:

```
child {
    start_action = trap
    rekey_time = 3600s
    esp_proposals = aes128-sha1-modp1024
    remote_ts = 10.240.0.0/24,100.94.242.0/24,10.54.108.0/24
    local_ts = 10.20.0.0/22
    dpd_action = restart
}
connections {
    IPsec-device-vpn-tunnel1 {
        rekey_time = 28800s
        version = 2
        local_addrs = x.x.x.x
        remote_addrs = z.z.z.z
        proposals = aes128-sha1-modp1024
        dpd_delay = 10s
        children {
            routing: child {
            }
        }
        local-0 {
                auth = psk
        }
        remote-0 {
                auth = psk
        }
    }
    IPsec-device-vpn-tunnel2 {
        rekey_time = 28800s
        version = 2
        local_addrs = x.x.x.x
        remote_addrs = y.y.y.y
        proposals = aes128-sha1-modp1024
        dpd_delay = 10s
        children {
            routing: child {
            }
        }
        local-0 {
                auth = psk
        }
        remote-0 {
                auth = psk
        }
    }
}
```
The version of strongswan is

```
$ swanctl --version
strongSwan swanctl 5.9.13
 ```
I collected the information since your reponses suggest, that it should not happen. I hope it gives some insights as to what is happening. If you need more or other logs, let me know.

**Reply by tobiasbrunner** [reply]
> I used `swanctl -T` to trace the log

That's generally not ideal to debug stuff as it's limited to debug level 1. Use the settings given in #196 instead to produce a debug log.

The log you posted shows that two SAs are created as **responder** (the configs are selected based on the remote IPs). If that's the case, it really doesn't matter if trap policies are used locally or not. The SAs can always be initiated by the peers. So you'll end up with two CHILD_SAs for the same traffic selectors. Only one will be used for outbound traffic (usually the last one installed, you could possibly control this by setting a manual priority). Inbound traffic could come via either SA.
